### DISCUSSION TITLE
How to pick scenegraph's parts?

### DISCUSSION THREAD

**Discussion by Kaszniam** [visgl/deck.gl#5428] [Ideas]
Hi!
Is there any smart way to be able to pick different parts of one scenegraph? For instance: let's imagine we have a fence built up from planks (each plank has its' own id and properties). This fence is loaded as scenegraph and it contains all info about planks inside its' 'children' property. Right now, the picking system treat it as one model (so no matter where we point our cursor, it returns data for overall object - fence). I'd like to change this behaviour and be able to distinguish which plank is being pointed out. So far the only idea I have is to somehow change  color picking system so it'll calculate picking color not for whole scenegraph loaded on the layer but rather for its children. Is it a good way? Or are there any simpler solutions?

**Comment by 29942016** [comment]
I'm currently trying to implement a similar feature for our scenegraphs. I don't think DeckGL has a native way to determine the scenegraph node you're hovering over (I'd love to be corrected here). I currently have two ideas for this:

1. Manually implementing ray casting; converting each scenegraph child node from 3d space into its 2d position using a world to screen function and then iterating which intersect with your cursors position.

2. Using another 3d library ontop of deckgl. ThreeJS includes a premade raycaster class. So we would need to copy the existing scenegraph + viewport into a ThreeJS scene and then use their raycaster intersection function. See: https://threejs.org/docs/#api/en/core/Raycaster

The second way is probably easier, but adds extra overhead.

Open to any suggestions or thoughts :)
